pkg/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,168 @@
+	Client        client.Interface
+	LeaseDuration time.Duration
+	RetryPeriod   time.Duration
+	OnAquire      func(stop <-chan struct{})
+	OnRelease     func()
+}
+
+type LeaderElectionRecord struct {
+	HolderIdentity       string
+	LeaseDurationSeconds int
+	AquireTime           unversioned.Time
+	RenewTime            unversioned.Time
+}
+
+func (le *LeaderElection) Run() {
+	defer func() {
+		util.HandleCrash()
+		le.OnRelease()
+	}()
+	le.aquire()
+	stop := make(chan struct{}, 1)
+	go le.OnAquire(stop)
+	le.renew()
+	stop <- struct{}{}
+}
+
+func (le *LeaderElection) aquire() {
+	stop := make(chan struct{}, 1)
+	util.Until(func() {
+		succeeded := le.tryAquireOrRenew()
+		if !succeeded {
+			glog.V(4).Infof("failed to renew lease %v/%v", le.EndpointsMeta.Namespace, le.EndpointsMeta.Name)
+			wait.Jitter(le.RetryPeriod, JitterFactor)
+			return
+		}
+		glog.Infof("sucessfully aquired lease %v/%v", le.EndpointsMeta.Namespace, le.EndpointsMeta.Name)
+		stop <- struct{}{}
+	}, 0, stop)
+}
+
+func (le *LeaderElection) renew() {
+	stop := make(chan struct{}, 1)
+	util.Until(func() {
+		err := wait.Poll(le.RetryPeriod, le.LeaseDuration, func() (bool, error) {
+			return le.tryAquireOrRenew(), nil
+		})
+		if err == nil {
+			glog.Infof("succesfully renewed lease %v/%v", le.EndpointsMeta.Namespace, le.EndpointsMeta.Name)
+			wait.Jitter(le.RetryPeriod, JitterFactor)
+			return
+		}
+		glog.Infof("failed to renew lease %v/%v", le.EndpointsMeta.Namespace, le.EndpointsMeta.Name)
+		stop <- struct{}{}
+	}, 0, stop)
+}
+
+func (le *LeaderElection) tryAquireOrRenew() bool {
+	now := unversioned.Now()
+	leaderElectionRecord := LeaderElectionRecord{
+		HolderIdentity:       le.Identity,
+		LeaseDurationSeconds: int(le.LeaseDuration / time.Second),
+		RenewTime:            now,
+		AquireTime:           now,
+	}
+
+	e, err := le.Client.Endpoints(le.EndpointsMeta.Namespace).Get(le.EndpointsMeta.Name)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return false
+		}
+
+		leaderElectionRecordBytes, err := json.Marshal(leaderElectionRecord)
+		if err != nil {
+			return false
+		}
+		_, err = le.Client.Endpoints(le.EndpointsMeta.Namespace).Create(&api.Endpoints{
+			ObjectMeta: api.ObjectMeta{
+				Name:      le.EndpointsMeta.Name,
+				Namespace: le.EndpointsMeta.Namespace,
+				Annotations: map[string]string{
+					LeaderElectionRecordAnnotationKey: string(leaderElectionRecordBytes),
+				},
+			},
+		})
+		if err != nil {
+			glog.Errorf("error initially creating endpoints: %v", err)
+			return false
+		}
+		return true
+	}
+
+	if oldLeaderElectionRecordBytes, found := e.Annotations[LeaderElectionRecordAnnotationKey]; found {
+		var oldLeaderElectionRecord LeaderElectionRecord
+		if err := json.Unmarshal([]byte(oldLeaderElectionRecordBytes), &oldLeaderElectionRecord); err != nil {
+			glog.Errorf("error unmarshaling leader election record: %v", err)
+			return false
+		}
+		if oldLeaderElectionRecord.RenewTime.Time.Add(le.LeaseDuration).After(now.Time) &&
```

Comment:
This could be made to use a locally captured timestamp but would be slightly more complicated